Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java

```diff
@@ -55,6 +89,39 @@
         return new ArrayDafnySequence<>(values);
     }
 
+    @SuppressWarnings("unchecked")
+    public T[] toArray(Class<T> type) {
```

Comment:
This could be a lot more efficient with the help of a `copyIntoArray(T[] array, int index)` helper method which could be overridden by the various subtypes to avoid the intermediate List<T> where possible (assuming that still makes sense after your subsequent PRs)

Reply:
Since `asList()` is just a view object, there isn't really an intermediate list (though of course bulk copying methods are always fast).  But yeah, this is something that changes again with the unboxed primitives PR (and there the subclasses do use bulk operations; CTTOI `toArray()` could probably be abstract in the superclass at that point).